### PR TITLE
[SPARK-43999][SQL][CORE] Support force finish useless stage when AQE on

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
This PR bring force finish stage feature to AQE, it used when some stage will be useless when plan after reOptimize.
eg:
```sql
SELECT * FROM emptyTestData t1 LEFT OUTER JOIN testData t2 ON t1.key = t2.key
```
The left table data is emtpy, so the right table data will be useless, unnecessary continue to read right table.
It will save database/spark CPU and IO resource if right table is large.
The Spark UI display force finish like
image

### Why are the changes needed?
add new force finish feature to AQE

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
add new test

### CODE REVIEW

**Comment by LuciferYang** [comment]
cc @cloud-fan @ulysses-you FYI

File: project/MimaExcludes.scala [hunk 1]
```diff
@@ -51,6 +51,8 @@ object MimaExcludes {
     ProblemFilters.exclude[ReversedMissingMethodProblem]("org.apache.spark.sql.catalog.Catalog.listTables"),
     // [SPARK-43992][SQL][PYTHON][CONNECT] Add optional pattern for Catalog.listFunctions
     ProblemFilters.exclude[ReversedMissingMethodProblem]("org.apache.spark.sql.catalog.Catalog.listFunctions"),
+    // [SPARK-43999][SQL] Support force finish useless stage when AQE on
```
Comment (LuciferYang) [inline project/MimaExcludes.scala:54]: really need this item?

Comment (Hisoka-X) [reply project/MimaExcludes.scala:54]: I consider remove it now, doesn't look elegant.

**Comment by Hisoka-X** [comment]
> hmm...If I remember correctly, @ulysses-you submitted a pr with similar goals

Thanks @LuciferYang , I think you are talk about https://github.com/apache/spark/pull/41536/ . I think there are some different between two PR. @ulysses-you 's way will affect reuse exchange (I'm not sure this PR will or not, but according the CI, seem will not), this PR will force finish stage not cancel it. The behavior are different.

By the way, please search `Unnecessary Stage` in https://github.com/Hisoka-X/spark/actions/runs/5387239085/jobs/9778311879 or https://github.com/Hisoka-X/spark/actions/runs/5387239085/jobs/9778311577 . Some query already benefited from this PR.

Anyway, I'm not sure this way are best, only convert my thought to code.

**Comment by ulysses-you** [comment]
The reason why I closed my pr is that we can not cancel a broadcast query stage because exchange reuse shares the broadcast future instance. It's ok to cancel a shuffle stage before executing final plan, but note that we can not cancel an intermediate shuffle stage to avoid breaking reuse shuffle exchange.

**Comment by ulysses-you** [comment]
The problem is that, we can not predicate which shuffle will be reused easily since AQE is adaptive. What if we cancel a running exchange but later we need to run another exchange which is semantic equivalent ?

**Pull request closed without merging**